File: src/algorithms/strings/kmp-search/kmp-search.test.ts
import { describe, it, expect } from "vitest";
import { kmpSearch } from "./sources/kmp-search.ts?fn";

describe("kmpSearch", () => {
  it("finds the pattern at the start of the text", () => {
    expect(kmpSearch("ABCDEF", "ABC")).toBe(0);
  });

  it("finds the pattern in the middle of the text", () => {
    expect(kmpSearch("ABABDABACDABABCABAB", "ABABCABAB")).toBe(10);
  });

  it("finds the pattern at the end of the text", () => {
    expect(kmpSearch("XYZABC", "ABC")).toBe(3);
  });

  it("returns -1 when the pattern is not present", () => {
    expect(kmpSearch("ABCDEFG", "XYZ")).toBe(-1);
  });

  it("handles a single-character pattern that exists", () => {
    expect(kmpSearch("HELLO", "L")).toBe(2);
  });

  it("handles a single-character pattern that does not exist", () => {
    expect(kmpSearch("HELLO", "Z")).toBe(-1);
  });

  it("returns 0 for an empty pattern", () => {
    expect(kmpSearch("HELLO", "")).toBe(0);
  });

  it("handles text equal to the pattern", () => {
    expect(kmpSearch("ABCD", "ABCD")).toBe(0);
  });

  it("returns -1 when pattern is longer than text", () => {
    expect(kmpSearch("AB", "ABCD")).toBe(-1);
  });

  it("handles repeated characters correctly", () => {
    expect(kmpSearch("AAAAAB", "AAAB")).toBe(2);
  });
});

File: src/algorithms/strings/pattern-matching/kmp-search/__tests__/kmp-search_test.go
package main

import "testing"

func TestKmpSearchPatternAtStart(t *testing.T) {
	if kmpSearch("ABCDEF", "ABC") != 0 {
		t.Error("expected 0")
	}
}

func TestKmpSearchPatternInMiddle(t *testing.T) {
	if kmpSearch("ABABDABACDABABCABAB", "ABABCABAB") != 10 {
		t.Error("expected 10")
	}
}

func TestKmpSearchPatternAtEnd(t *testing.T) {
	if kmpSearch("XYZABC", "ABC") != 3 {
		t.Error("expected 3")
	}
}

func TestKmpSearchPatternNotFound(t *testing.T) {
	if kmpSearch("ABCDEFG", "XYZ") != -1 {
		t.Error("expected -1")
	}
}

func TestKmpSearchSingleCharFound(t *testing.T) {
	if kmpSearch("HELLO", "L") != 2 {
		t.Error("expected 2")
	}
}

func TestKmpSearchSingleCharNotFound(t *testing.T) {
	if kmpSearch("HELLO", "Z") != -1 {
		t.Error("expected -1")
	}
}

func TestKmpSearchEmptyPattern(t *testing.T) {
	if kmpSearch("HELLO", "") != 0 {
		t.Error("expected 0 for empty pattern")
	}
}

func TestKmpSearchTextEqualsPattern(t *testing.T) {
	if kmpSearch("ABCD", "ABCD") != 0 {
		t.Error("expected 0")
	}
}

func TestKmpSearchPatternLongerThanText(t *testing.T) {
	if kmpSearch("AB", "ABCD") != -1 {
		t.Error("expected -1")
	}
}

func TestKmpSearchRepeatedCharacters(t *testing.T) {
	if kmpSearch("AAAAAB", "AAAB") != 2 {
		t.Error("expected 2")
	}
}
